README.md
# Browserslist Useragent

![build](https://github.com/pastelsky/browserslist-useragent/actions/workflows/main.yml/badge.svg)
[![npm](https://img.shields.io/npm/v/browserslist-useragent.svg)](https://www.npmjs.com/package/browserslist-useragent)

<img align="right" width="120" height="120"
     src="https://cdn.rawgit.com/pastelsky/browserslist-useragent/master/logo.svg" alt="Browserslist Useragent logo (original by Anton Lovchikov)" />

Find if a given user agent string satisfies a [browserslist](https://github.com/ai/browserslist) query.

It automatically reads the browserslist configuration specified in your project,
but you can also specify the same using the `options` parameter.

**If you wish to target modern browsers, read [this](#when-querying-for-modern-browsers).**

## Installation

```bash
npm install browserslist-useragent
```

## Usage

```js
const { matchesUA } = require('browserslist-useragent')

matchesUA(userAgentString, options)

// with browserslist config inferred
matchesUA('Mozilla/5.0 (Windows NT 10.0; rv:54.0) Gecko/20100101 Firefox/54.0')
//returns boolean

// with explicit browserslist
matchesUA(
  'Mozilla/5.0 (Windows NT 10.0; rv:54.0) Gecko/20100101 Firefox/54.0',
  { browsers: ['Firefox > 53'] }
)
// returns true
```

| Option              | Default Value   | Description                                                                                                                                                                                                                |
| ------------------- | --------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| browsers            | —               | Manually provide a browserslist query (or an array of queries). Specifying this overrides the browserslist configuration specified in your project.                                                                        |
| env                 | —               | When multiple browserslist [environments](https://github.com/ai/browserslist#environments) are specified, pick the config belonging to this environment.                                                                   |
| path                | `process.cwd()` | Specify a folder to search for the browserslist config (if it differs from the current working directory)                                                                                                                  |
| ignorePatch         | `true`          | Ignore differences in patch browser numbers                                                                                                                                                                                |
| ignoreMinor         | `false`         | Ignore differences in minor browser versions                                                                                                                                                                               |
| allowHigherVersions | `false`         | For all the browsers in the browserslist query, return a match if the user agent version is equal to or higher than the one specified in browserslist. See [why](#when-querying-for-modern-browsers) this might be useful. |

## Supported browsers

- Chrome (Chrome / Chromium / Yandex) as `and_chr` | `ChromeAndroid` | `Chrome`
- Samsung Internet as `Samsung`
- Firefox as `ff` | `and_ff` | `FirefoxAndroid` | `Firefox`
- Safari iOS as `ios_saf` | `iOS`
- Safari Desktop as `Safari`
- IE as `ie` | `ie_mob`
- Edge as `Edge`
- Electron as `Electron`

PRs to add more _browserslist supported_ browsers are welcome 👋

## Notes

- All browsers on iOS (Chrome, Firefox etc) use Safari's WebKit as the underlying engine, and hence will be resolved to Safari. Since `browserslist` is usually used for
  transpiling / autoprefixing for browsers, this behaviour is what's intended in most cases, but might surprise you otherwise.

- Right now, Chrome for Android and Firefox for Android are resolved to their desktop equivalents. The `caniuse` database does not currently store historical data for these browsers separately (except the last version) See [#156](https://github.com/ai/browserslist/issues/156). However,
  safari for iOS and desktop can be matched separately, since this data is available for both.

## When querying for modern browsers

- It is a good idea to update this package often so that browser definitions are upto date.
- It is also a good idea to add `unreleased versions` to your browserslist query, and set `ignoreMinor` and `ignorePatch` to true so that alpha / beta / canary versions of browsers are matched.
- In case you're unable to keep this package up-to-date, you can set the `allowHigherVersions` to `true`. For all the browsers specified in your browserslist query, this will return a match if the user agent version is equal to or higher than those specified in your browserslist query. Use this with care though, since it's a wildcard, and only lightly tested.

## Further reads

- [Smart Bundling: Shipping legacy code to only legacy browsers](https://www.smashingmagazine.com/2018/10/smart-bundling-legacy-code-browsers/)


CHANGELOG.md
Please use [github releases page](https://github.com/browserslist/browserslist-useragent/releases) to see changes for future versions.

# v3.0.0

- **Fix**: Don't crash when `options` parameter is not specified
- **Fix**: Use `loose` option to parse browsers versions. Doesn't crash on versions like `6.00.1` or 6.0.0005` etc.

- **Feature**: Added support for Samsung Internet

- **Breaking**: Safari on desktop and on iOS are now considered as separate browsers.
  You'll need to use separate browserslist queries to target both of them - `['Safari >= 10', 'iOS >= 10.1']`. See #23
- **Breaking**: Removing support for undocumented option to `matchesUA` - `all`.
  This was used internally, and shouldn't be breaking change for you unless you poked around code to discover it.
- **Breaking**: Remove deprecated `_allowHigherVersions` flag. This has been replaced by `allowHigherVersions` in the past.

# v2.0.1

- Remove deprecated \_allowHigherVersions option from README

# v2.0.0

- **Feature**: Added Yandex to list of supported browsers
- **Feature**: Added support for Yandex Browser
- **Deprecation**: Deprecated `_allowHigherVersions`

#v1.2.1

- Fixed build link

#v1.2.0

- **Feature**: Updated packages and added support for headless chrome and chromium
- Fixed links in badges

#v1.1.0

- **Feature**: Added support for selecting environment
- Updated browserslist version

#v1.0.4

- delete gitignore

#v1.0.3

- Update README.md

#v1.0.2

- Second try at fixing logo

#v1.0.1

- Updated browserslist version

#v1.0.0

- Create LICENSE
- /browserlist/ => browserslist
- Refactored code and rewrote Readme

#v0.0.2

- Added install instructions
- Initial commit


tests/index.test.js
const ua = require('useragent-generator')

const { resolveUserAgent, matchesUA, normalizeQuery } = require('../index')

const CustomUserAgentString = {
  YANDEX:
    'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 YaBrowser/18.1.1.839 Yowser/2.5 Safari/537.36',
  YANDEX_SEARCH:
    'Mozilla/5.0 (Linux; Android 4.4.4; GT-I9300I Build/KTU84P) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/33.0.0.0 Mobile Safari/537.36 YandexSearch/7.16',
  SAMSUNG_BROWSER_7_2:
    'Mozilla/5.0 (Linux; Android 6.0.1; SAMSUNG SM-N930F Build/MMB29K) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/7.2 Chrome/63.0.3239.111 Mobile Safari/537.36',
  FACEBOOK_WEBVIEW_CHROME_ANDROID:
    'Mozilla/5.0 (Linux; Android 8.0.0; LG-H930 Build/OPR1.170623.026; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36 [FB_IAB/Orca-Android;FBAV/189.0.0.27.99;]',
  FACEBOOK_WEBVIEW_IOS:
    'Mozilla/5.0 (iPhone; CPU iPhone OS 8_2 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Mobile/12D508 [FBAN/FBIOS;FBAV/27.0.0.10.12;FBBV/8291884;FBDV/iPhone7,1;FBMD/iPhone;FBSN/iPhone OS;FBSV/8.2;FBSS/3; FBCR/vodafoneIE;FBID/phone;FBLC/en_US;FBOP/5]',
  OPERA_MOBILE:
    'Opera/9.80 (Android 2.3.3; Linux; Opera Mobi/ADR-1111101157; U; es-ES) Presto/2.9.201 Version/11.50',
  ELECTRON:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 11_4_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Electron/12.0.13 Safari/537.36',
}

it('normalizes queries properly', () => {
  expect(normalizeQuery('and_chr >= 61')).toBe('Chrome >= 61')

  expect(normalizeQuery('ChromeAndroid >= 61')).toBe('Chrome >= 61')

  expect(normalizeQuery('FirefoxAndroid < 54')).toBe('Firefox < 54')

  expect(normalizeQuery('ff < 54')).toBe('Firefox < 54')

  expect(normalizeQuery('ios_saf < 10.1.0')).toBe('iOS < 10.1.0')

  expect(normalizeQuery('last 10 and_chr versions')).toBe(
    'last 10 Chrome versions'
  )

  expect(normalizeQuery('>= 5%')).toBe('>= 5%')
})

it('resolves all browsers in iOS to safari with correct platform version', () => {
  expect(resolveUserAgent(ua.chrome.iOS('10.3.0'))).toEqual({
    family: 'iOS',
    version: '10.3.0',
  })

  expect(resolveUserAgent(ua.firefox.iOS('10.3.0'))).toEqual({
    family: 'iOS',
    version: '10.3.0',
  })

  expect(resolveUserAgent(ua.safari.iOSWebview('10.3.0'))).toEqual({
    family: 'iOS',
    version: '10.3.0',
  })

  expect(resolveUserAgent(ua.safari.iOS('8.3.0'))).toEqual({
    family: 'iOS',
    version: '8.3.0',
  })
})

it('resolves desktop safari on osx properly', () => {
  expect(resolveUserAgent(ua.safari('10.1.0'))).toEqual({
    family: 'Safari',
    version: '10.1.0',
  })
})

it('resolves IE/Edge properly', () => {
  expect(resolveUserAgent(ua.ie('11.0.0'))).toEqual({
    family: 'Explorer',
    version: '11.0.0',
  })

  expect(resolveUserAgent(ua.ie.windowsPhone('10.0.0'))).toEqual({
    family: 'ExplorerMobile',
    version: '10.0.0',
  })

  expect(resolveUserAgent(ua.edge('14.1.0'))).toEqual({
    family: 'Edge',
    version: '14.1.0',
  })
})

it('resolves chrome/android properly', () => {
  expect(resolveUserAgent(ua.chrome('41.0.228.90'))).toEqual({
    family: 'Chrome',
    version: '41.0.228',
  })

  expect(resolveUserAgent(ua.chrome.androidWebview('2.3.3'))).toEqual({
    family: 'Android',
    version: '2.3.3',
  })

  expect(
    resolveUserAgent(
      ua.chrome.androidWebview({
        androidVersion: '4.4.1',
        chromeVersion: '44.0.0',
      })
    )
  ).toEqual({
    family: 'Chrome',
    version: '44.0.0',
  })

  expect(
    resolveUserAgent(
      ua.chrome.androidWebview({
        androidVersion: '6.0.0',
        chromeVersion: '60.0.0',
      })
    )
  ).toEqual({
    family: 'Chrome',
    version: '60.0.0',
  })

  expect(
    resolveUserAgent(
      ua.chrome('41.0.228.90').replace('Chrome', 'HeadlessChrome')
    )
  ).toEqual({
    family: 'Chrome',
    version: '41.0.228',
  })

  expect(resolveUserAgent(ua.chromium('41.0.228.90'))).toEqual({
    family: 'Chrome',
    version: '41.0.228',
  })

  expect(
    matchesUA(CustomUserAgentString.YANDEX, { browsers: ['Chrome >= 63'] })
  ).toBeTruthy()

  expect(
    matchesUA(CustomUserAgentString.YANDEX_SEARCH, {
      browsers: ['Chrome >= 33'],
    })
  ).toBeTruthy()

  expect(
    resolveUserAgent(CustomUserAgentString.FACEBOOK_WEBVIEW_CHROME_ANDROID)
  ).toEqual({
    family: 'Chrome',
    version: '69.0.3497',
  })

  expect(
    matchesUA(CustomUserAgentString.FACEBOOK_WEBVIEW_CHROME_ANDROID, {
      browsers: ['Chrome >= 63'],
    })
  ).toBeTruthy()

  expect(resolveUserAgent(CustomUserAgentString.FACEBOOK_WEBVIEW_IOS)).toEqual({
    family: 'iOS',
    version: '8.2.0',
  })

  expect(
    matchesUA(CustomUserAgentString.FACEBOOK_WEBVIEW_IOS, {
      browsers: ['iOS >= 8'],
    })
  ).toBeTruthy()
})

it('resolves firefox properly', () => {
  expect(resolveUserAgent(ua.firefox('41.0.0'))).toEqual({
    family: 'Firefox',
    version: '41.0.0',
  })

  expect(resolveUserAgent(ua.firefox.androidPhone('44.0.0'))).toEqual({
    family: 'Firefox',
    version: '44.0.0',
  })
})

it('resolves samsung browser properly', () => {
  expect(resolveUserAgent(CustomUserAgentString.SAMSUNG_BROWSER_7_2)).toEqual({
    family: 'Samsung',
    version: '7.2.0',
  })
})

it('resolves electron properly', () => {
  // Electron 12 -> Chrome 89
  expect(resolveUserAgent(CustomUserAgentString.ELECTRON)).toEqual({
    family: 'Chrome',
    version: '89',
  })
})

it('detects if browserslist matches UA', () => {
  expect(
    matchesUA(ua.firefox.androidPhone('40.0.1'), {
      browsers: ['Firefox >= 40'],
    })
  ).toBeTruthy()

  expect(
    matchesUA(ua.firefox('30.0.0'), { browsers: ['Firefox >= 10.0.0'] })
  ).toBeTruthy()

  expect(
    matchesUA(ua.chrome.iOS('11.0.0'), { browsers: ['iOS >= 10.3.0'] })
  ).toBeTruthy()

  expect(
    matchesUA(ua.safari.iOS('11.0.0'), { browsers: ['iOS >= 10.3.0'] })
  ).toBeTruthy()

  expect(
    matchesUA(CustomUserAgentString.SAMSUNG_BROWSER_7_2, {
      browsers: ['Samsung >= 7'],
    })
  ).toBeTruthy()

  expect(
    matchesUA(CustomUserAgentString.OPERA_MOBILE, {
      browsers: ['OperaMobile >= 7'],
      ignoreMinor: true,
    })
  ).toBeTruthy()

  const modernList = [
    'Firefox >= 53',
    'Edge >= 15',
    'Chrome >= 58',
    'iOS >= 10',
    'Safari >= 11',
    'Samsung >= 7',
  ]

  expect(matchesUA(ua.safari.iOS(9), { browsers: modernList })).toBeFalsy()

  expect(
    matchesUA(ua.chrome.androidPhone(57), { browsers: modernList })
  ).toBeFalsy()

  expect(
    matchesUA(ua.firefox.androidPhone(52), { browsers: modernList })
  ).toBeFalsy()

  expect(matchesUA(ua.firefox(56), { browsers: modernList })).toBeTruthy()

  expect(matchesUA(ua.edge(14), { browsers: modernList })).toBeFalsy()

  expect(matchesUA(ua.chrome(64), { browsers: modernList })).toBeTruthy()

  expect(
    matchesUA(ua.chrome.androidWebview('4.3.3'), { browsers: modernList })
  ).toBeFalsy()

  expect(matchesUA(ua.safari('12.0.0'), { browsers: modernList })).toBeTruthy()

  expect(matchesUA(ua.safari('10.0.0'), { browsers: modernList })).toBeFalsy()

  expect(
    matchesUA(CustomUserAgentString.SAMSUNG_BROWSER_7_2, {
      browsers: modernList,
    })
  ).toBeTruthy()
})

it('can interpret various variations in specifying browser names', () => {
  expect(matchesUA(ua.chrome(49), { browsers: ['and_chr >= 49'] })).toBeTruthy()

  expect(
    matchesUA(ua.safari.iOS('10.3.0'), { browsers: ['ios_saf >= 10.1.0'] })
  ).toBeTruthy()

  expect(
    matchesUA(ua.firefox.androidPhone('46.0.0'), {
      browsers: ['FirefoxAndroid >= 41.1.0'],
    })
  ).toBeTruthy()
})

it('ignorePatch option works correctly', () => {
  expect(
    matchesUA(ua.firefox('49.0.1'), {
      browsers: ['ff >= 44'],
      ignorePatch: false,
    })
  ).toBeFalsy()

  expect(
    matchesUA(ua.firefox('49.0.1'), {
      browsers: ['ff >= 44'],
      ignorePatch: true,
    })
  ).toBeTruthy()

  expect(
    matchesUA(ua.firefox('49.1.1'), {
      browsers: ['ff >= 44'],
      ignorePatch: true,
      ignoreMinor: false,
    })
  ).toBeFalsy()
})

it('ignoreMinor option works correctly', () => {
  expect(
    matchesUA(ua.firefox('49.1.0'), {
      browsers: ['ff >= 44'],
      ignoreMinor: false,
    })
  ).toBeFalsy()

  expect(
    matchesUA(ua.firefox('49.1.0'), {
      browsers: ['ff >= 44'],
      ignoreMinor: true,
    })
  ).toBeTruthy()

  expect(
    matchesUA(ua.firefox('49.1.3'), {
      browsers: ['ff >= 44'],
      ignoreMinor: true,
      ignorePatch: false,
    })
  ).toBeTruthy()
})

it('allowHigherVersions works correctly', () => {
  expect(
    matchesUA(ua.chrome('99'), {
      browsers: ['chrome >= 60'],
      allowHigherVersions: false,
    })
  ).toBeFalsy()

  expect(
    matchesUA(ua.chrome('66'), {
      browsers: ['chrome >= 60'],
      allowHigherVersions: true,
    })
  ).toBeTruthy()
})

it('parses semvers liberally', () => {
  expect(
    matchesUA(
      'Opera/9.80 (Windows NT 6.1; U; es-ES) Presto/2.9.181 Version/12.00',
      { browsers: ['opera >= 12'] }
    )
  ).toBeTruthy()

  expect(
    matchesUA(
      'Opera/9.80 (Windows NT 6.1; U; es-ES) Presto/2.9.181 Version/12.000.1',
      { browsers: ['opera >= 12'] }
    )
  ).toBeTruthy()

  expect(
    matchesUA(
      'Opera/9.80 (Windows NT 6.1; U; es-ES) Presto/2.9.181 Version/12.0.0001',
      { browsers: ['opera >= 12'] }
    )
  ).toBeTruthy()

  expect(
    matchesUA(
      'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.9.1b2pre) Gecko/20081015 Fennec/55.0a1',
      { browsers: ['firefox > 50'], allowHigherVersions: false }
    )
  ).toBeTruthy()
})

it('can deal with version ranges (if returned by browserslist)', () => {
  expect(
    matchesUA(ua.safari.iOS('9.1.0'), { browsers: ['ios_saf >= 9'] })
  ).toBeTruthy()

  expect(
    matchesUA(ua.safari.iOS('9.0.0'), { browsers: ['ios_saf >= 9'] })
  ).toBeTruthy()

  // This should fail
  // see https://github.com/browserslist/browserslist/issues/402
  expect(
    matchesUA(ua.safari.iOS('9.1.0'), { browsers: ['ios_saf >= 9.2'] })
  ).toBeTruthy() // <-- should actually be falsy
})

it('can deal with non-numerical version numbers returned by browserslist for safari technology preview', () => {
  expect(
    matchesUA(ua.safari('18.1.0'), {
      browsers: ['unreleased Safari versions'],
      ignorePatch: true,
      ignoreMinor: true,
      allowHigherVersions: true,
    })
  ).toBeTruthy()
})

it('gracefully fails on invalid inputs', () => {
  expect(matchesUA(undefined)).toBeFalsy()

  expect(matchesUA(null)).toBeFalsy()
})


index.js
const browserslist = require('browserslist')
const semver = require('semver')
const useragent = require('useragent')
const e2c = require('electron-to-chromium/versions')

// @see https://github.com/ai/browserslist#browsers

// map of equivalent browsers,
// see https://github.com/ai/browserslist/issues/156

const browserNameMap = {
  bb: 'BlackBerry',
  and_chr: 'Chrome',
  ChromeAndroid: 'Chrome',
  FirefoxAndroid: 'Firefox',
  ff: 'Firefox',
  ie_mob: 'ExplorerMobile',
  ie: 'Explorer',
  and_ff: 'Firefox',
  ios_saf: 'iOS',
  op_mini: 'OperaMini',
  op_mob: 'OperaMobile',
  and_qq: 'QQAndroid',
  and_uc: 'UCAndroid',
}

function resolveUserAgent(uaString) {
  // Chrome and Opera on iOS uses a UIWebView of the underlying platform to render
  // content, by stripping the CriOS or OPiOS strings the useragent parser will alias the
  // user agent to ios_saf for the UIWebView, which is closer to the actual
  // renderer
  // @see https://github.com/Financial-Times/polyfill-service/pull/416

  let strippedUA = uaString.replace(
    /((CriOS|OPiOS)\/(\d+)\.(\d+)\.(\d+)\.(\d+))/,
    ''
  )

  // Yandex Browser uses Chromium as the udnerlying engine
  strippedUA = strippedUA.replace(/YaBrowser\/(\d+\.?)+/g, '')

  // Yandex Search uses Chromium as the udnerlying engine
  strippedUA = strippedUA.replace(/YandexSearch\/(\d+\.?)+/g, '')

  // Facebook Webview
  strippedUA = strippedUA.replace(/FB_IAB/g, '').replace(/FBAN\/FBIOS/g, '')

  const parsedUA = useragent.parse(strippedUA)

  // Case A: For Safari, Chrome and others browsers on iOS
  // that report as Safari after stripping tags
  if (parsedUA.family.includes('Safari') && parsedUA.os.family === 'iOS') {
    return {
      family: 'iOS',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  // Case B: The browser on iOS didn't report as safari,
  // so we use the iOS version as a proxy to the browser
  // version. This is based on the assumption that the
  // underlying Safari Engine used will be *atleast* equal
  // to the iOS version it's running on.
  if (parsedUA.os.family === 'iOS') {
    return {
      family: 'iOS',
      version: [parsedUA.os.major, parsedUA.os.minor, parsedUA.os.patch].join(
        '.'
      ),
    }
  }

  // Case C: The caniuse database does not contain
  // historical browser versions for so called `minor`
  // browsers like Chrome for Android, Firefox for Android etc
  // In this case, we proxy to the desktop version
  // @see https://github.com/Fyrd/caniuse/issues/3518

  if (
    parsedUA.family.includes('Chrome Mobile') ||
    parsedUA.family.includes('Chrome Mobile WebView') ||
    parsedUA.family.includes('Chromium') ||
    parsedUA.family.includes('HeadlessChrome')
  ) {
    return {
      family: 'Chrome',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'Opera Mobile') {
    return {
      family: 'OperaMobile',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'Samsung Internet') {
    return {
      family: 'Samsung',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'Firefox Mobile') {
    return {
      family: 'Firefox',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'IE') {
    return {
      family: 'Explorer',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'IE Mobile') {
    return {
      family: 'ExplorerMobile',
      version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
    }
  }

  if (parsedUA.family === 'Electron') {
    const electronVersion = [parsedUA.major, parsedUA.minor].join('.')
    return {
      family: 'Chrome',
      version: e2c[electronVersion],
    }
  }

  return {
    family: parsedUA.family,
    version: [parsedUA.major, parsedUA.minor, parsedUA.patch].join('.'),
  }
}

// Convert version to a semver value.
// 2.5 -> 2.5.0; 1 -> 1.0.0;
const semverify = (version) => semver.coerce(version, { loose: true }).version

function flatten(arr) {
  return arr.reduce(function (flat, toFlatten) {
    return flat.concat(
      Array.isArray(toFlatten) ? flatten(toFlatten) : toFlatten
    )
  }, [])
}

// 10.0-10.2 -> 10.0, 10.1, 10.2
function generateSemversInRange(versionRange) {
  const [start, end] = versionRange.split('-')
  const startSemver = semverify(start)
  const endSemver = semverify(end)
  const versionsInRange = []
  let curVersion = startSemver

  while (semver.gte(endSemver, curVersion)) {
    versionsInRange.push(curVersion)
    curVersion = semver.inc(curVersion, 'minor')
  }

  return versionsInRange
}

function normalizeQuery(query) {
  let normalizedQuery = query
  const regex = `(${Object.keys(browserNameMap).join('|')})`
  const match = query.match(new RegExp(regex))

  if (match) {
    normalizedQuery = query.replace(match[0], browserNameMap[match[0]])
  }

  return normalizedQuery
}

const parseBrowsersList = (browsersList) => {
  const browsers = browsersList
    .map((browser) => {
      const [name, version] = browser.split(' ')
      return { name, version }
    })
    // #38 Filter out non-numerical browser versions
    .filter((browser) => browser.version !== 'TP')
    .map((browser) => {
      let normalizedName = browser.name
      let normalizedVersion = browser.version

      if (browser.name in browserNameMap) {
        normalizedName = browserNameMap[browser.name]
      }

      // browserslist might return ranges (9.0-9.2), unwrap them
      // see https://github.com/browserslist/browserslist-useragent/issues/41
      if (browser.version.indexOf('-') > 0) {
        return generateSemversInRange(browser.version).map((version) => ({
          family: normalizedName,
          version,
        }))
      } else {
        return {
          family: normalizedName,
          version: normalizedVersion,
        }
      }
    })

  return flatten(browsers)
}

const compareBrowserSemvers = (versionA, versionB, options) => {
  const semverifiedA = semverify(versionA)
  const semverifiedB = semverify(versionB)
  let referenceVersion = semverifiedB

  if (options.ignorePatch) {
    referenceVersion = `~${semverifiedB}`
  }

  if (options.ignoreMinor) {
    referenceVersion = `^${semverifiedB}`
  }

  if (options.allowHigherVersions) {
    if (semverifiedB === "TP.0.0") {
      return semverifiedA === semverifiedB;
    }
    return semver.gte(semverifiedA, semverifiedB)
  } else {
    return semver.satisfies(semverifiedA, referenceVersion)
  }
}

const matchesUA = (uaString, opts = {}) => {
  // bail out early if the user agent is invalid
  if (!uaString) {
    return false
  }

  let normalizedQuery
  if (opts.browsers) {
    normalizedQuery = opts.browsers.map(normalizeQuery)
  }
  const browsers = browserslist(normalizedQuery, {
    env: opts.env,
    path: opts.path || process.cwd(),
  })

  const parsedBrowsers = parseBrowsersList(browsers)

  const resolvedUserAgent = resolveUserAgent(uaString)

  const options = {
    ignoreMinor: false,
    ignorePatch: true,
    ...opts,
  }

  return parsedBrowsers.some((browser) => {
    return (
      browser.family.toLowerCase() ===
        resolvedUserAgent.family.toLocaleLowerCase() &&
      compareBrowserSemvers(resolvedUserAgent.version, browser.version, options)
    )
  })
}

module.exports = {
  matchesUA,
  resolveUserAgent,
  normalizeQuery,
}
